### PR TITLE
fix keep order in table scan is lost when remote read of PartitionTableScan(release-7.1)

### LINKED ISSUE #7519: running ch partition workload and meets error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. run ch list partition workload
2. check consistency between tiflash and tikv

### 3. What did you see instead (Required)
[2023/05/21 23:25:24.121 +08:00] [ERROR] [ch.go:256] ["check warehouse 2 at condition [3.3.2.5](http://3.3.2.5/) failed exec SELECT count(*)  FROM orders LEFT
 JOIN new_order ON (no_w_id=o_w_id AND o_d_id=no_d_id AND o_id=no_o_id) where o_w_id = ? and ((o_carrier_id IS NULL and no_o_id IS  NULL) O
R (o_carrier_id IS NOT NULL and no_o_id IS NOT NULL  ))  failed Error 1105: other error for mpp stream: From MPP<query:<query_ts:1684682723
943347087, local_query_id:77237, server_id:2001687, start_ts:441629467975614567>,task_id:4>: Code: 0, e.displayText() = DB::Exception: Code
: 0, e.displayText() = DB::Exception: code: 49\nmsg: \"Check !filter || !filter->before_where || read_mode == ReadMode::Bitmap failed: Push
 down filters needs bitmap: (while creating InputStreams from storage `db_31987`.`t_34126`, table_id: 34126)\"\n, e.what() = DB::Exception,
"] [stack="[github.com/pingcap/endless/testcase/tiflash/pkg/workload.(*ChWorkload).checkConsistence\n\t/home/jenkins/agent/workspace/endless](http://github.com/pingcap/endless/testcase/tiflash/pkg/workload.(*ChWorkload).checkConsistence/n/t/home/jenkins/agent/workspace/endless)
-master-build/testcase/tiflash/pkg/workload/ch.go:256\[ngithub.com/pingcap/endless/testcase/tiflash/pkg/workload.(*ChWorkload).RunTP.func3\n](http://ngithub.com/pingcap/endless/testcase/tiflash/pkg/workload.(*ChWorkload).RunTP.func3/n)
\t/home/jenkins/agent/workspace/endless-master-build/testcase/tiflash/pkg/workload/ch.go:208\[ngolang.org/x/sync/errgroup.(*Group).Go.func1](http://ngolang.org/x/sync/errgroup.(*Group).Go.func1)\
n\t/go/pkg/mod/[golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75](http://golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75)"]
[2023/05/21 23:27:51.519 +08:00] [ERROR] [ch.go:256] ["check warehouse 7 at condition 3.3.2.10 failed exec SELECT count(*) \n\tFROM (  SELE
CT  c.c_id, c.c_d_id, c.c_w_id, c.c_balance c1, \n\t\t\t\t   (SELECT sum(ol_amount) FROM orders STRAIGHT_JOIN order_line \n\t\t\t\t\t WHERE
 OL_W_ID=O_W_ID \n\t\t\t\t\t   AND OL_D_ID = O_D_ID \n\t\t\t\t\t   AND OL_O_ID = O_ID \n\t\t\t\t\t   AND OL_DELIVERY_D IS NOT NULL \n\t\t\t
\t\t   AND O_W_ID=? \n\t\t\t\t\t   AND O_D_ID=c.C_D_ID \n\t\t\t\t\t   AND O_C_ID=c.C_ID) sm, (SELECT  sum(h_amount)  from  history \n\t\t\t
\t\t\t\t\t\t\t\t\tWHERE H_C_W_ID=? \n\t\t\t\t\t\t\t\t\t\t\t\t  AND H_C_D_ID=c.C_D_ID \n\t\t\t\t\t\t\t\t\t\t\t\t  AND H_C_ID=c.C_ID) smh \n\
t\t\t FROM customer c \n\t\t\tWHERE  c.c_w_id = ? ) t\n   WHERE c1<>sm-smh failed Error 1105: other error for mpp stream: From MPP<query:<q
uery_ts:1684682861658032567, local_query_id:77264, server_id:2001687, start_ts:441629504072843366>,task_id:10>: Code: 0, e.displayText() =
DB::Exception: Code: 0, e.displayText() = DB::Exception: code: 49\nmsg: \"Check !filter || !filter->before_where || read_mode == ReadMode::
Bitmap failed: Push down filters needs bitmap: (while creating InputStreams from storage `db_31987`.`t_34636`, table_id: 34636)\"\n, e.what
() = DB::Exception,"] [stack="github.com/pingcap/endless/testcase/tiflash/pkg/workload.(*ChWorkload).checkConsistence\n\t/home/jenkins/agen
t/workspace/endless-master-build/testcase/tiflash/pkg/workload/ch.go:256\ngithub.com/pingcap/endless/testcase/tiflash/pkg/workload.(*ChWork
load).RunTP.func3\n\t/home/jenkins/agent/workspace/endless-master-build/testcase/tiflash/pkg/workload/ch.go:208\ngolang.org/x/sync/errgroup
.(*Group).Go.func1\n\t/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75"]
### 4. What is your TiFlash version? (Required)
sh-5.1# ./tiflash --version
TiFlash
Release Version: v7.1.0
Edition:         Community
Git Commit Hash: 70200bcd6d32f29faa9073a3e0a19c849abef241
Git Branch:      heads/refs/tags/v7.1.0
UTC Build Time:  2023-05-12 23:23:30
Enable Features: jemalloc sm4(GmSSL) avx2 avx512 unwind thinlto
Profile:         RELWITHDEBINFO

Raft Proxy
Git Commit Hash:   5cdf47726c58e04683c2da3e2892c276fccb3e8a
Git Commit Branch: HEAD
UTC Build Time:    2023-05-12 23:32:47
Rust Version:      rustc 1.67.0-nightly (96ddd32c4 2022-11-14)
Storage Engine:    tiflash
Prometheus Prefix: tiflash_proxy_
Profile:           release
Enable Features:   Unknown (env var does not exist when building)

### LINKED ISSUE #7527: fast scan does not take effect when partition scan
When it is partition scan and set fast scan to true, tiflash will still return the result of normal scan.

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
Signed-off-by: Lloyd-Pottiger <yan1579196623@gmail.com>### What problem does this PR solve?

Issue Number: close #7519, close #7527

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by SeaRise — APPROVED** [review]
Needs more investigation on master branch

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp [hunk 1]
```diff
@@ -1050,11 +1053,13 @@ BlockInputStreams DeltaMergeStore::read(const Context & db_context,
     SegmentReadTasks tasks = getReadTasksByRanges(*dm_context, sorted_ranges, num_streams, read_segments, /*try_split_task =*/!enable_read_thread);
     auto log_tracing_id = getLogTracingId(*dm_context);
     auto tracing_logger = log->getChild(log_tracing_id);
-    LOG_DEBUG(tracing_logger,
-              "Read create segment snapshot done, keep_order={} dt_enable_read_thread={} enable_read_thread={}",
-              keep_order,
-              db_context.getSettingsRef().dt_enable_read_thread,
-              enable_read_thread);
+    LOG_INFO(tracing_logger,
+             "Read create segment snapshot done, keep_order={} dt_enable_read_thread={} enable_read_thread={}, is_fast_scan={}, is_push_down_filter_empty={}",
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp:1057]:
```suggestion
             "Read create segment snapshot done, keep_order={} dt_enable_read_thread={} enable_read_thread={} is_fast_scan={} is_push_down_filter_empty={}",
```

File: dbms/src/Flash/Coprocessor/TiDBTableScan.cpp [hunk 2]
```diff
@@ -28,8 +28,10 @@ TiDBTableScan::TiDBTableScan(
     , pushed_down_filters(is_partition_table_scan ? std::move(table_scan->partition_table_scan().pushed_down_filter_conditions()) : std::move(table_scan->tbl_scan().pushed_down_filter_conditions()))
     // Only No-partition table need keep order when tablescan executor required keep order.
     // If keep_order is not set, keep order for safety.
-    , keep_order(!is_partition_table_scan && (table_scan->tbl_scan().keep_order() || !table_scan->tbl_scan().has_keep_order()))
-    , is_fast_scan(table_scan->tbl_scan().is_fast_scan())
+    // When keep_order is true, we will not push down filters to tablescan in TiDB.
+    // So even if keep_order is not set (may be lost), but pushed down filters is not empty, we can't keep order.
+    , keep_order(!is_partition_table_scan && pushed_down_filters.empty() && (!table_scan->tbl_scan().has_keep_order() || table_scan->tbl_scan().keep_order()))
```
Comment (JaySon-Huang) [inline dbms/src/Flash/Coprocessor/TiDBTableScan.cpp:33]: We should not do this hack. It is hard to understand and maintain.
Instead, an explicit check and throw a TiFlashException(Errors::Coprocessor::BadRequest, ...) is better

Comment (Lloyd-Pottiger) [reply dbms/src/Flash/Coprocessor/TiDBTableScan.cpp:33]: ok

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**